Grandfather in c10d_functional ops to pt2_compliant

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #113061
* #113050
* __->__ #113049

This PR also adds the ability to specify Tags for more `m.def(`
overloads.

Test Plan:
- new test

zou3519 committed:
Grandfather in c10d_functional ops to pt2_compliant

This PR also adds the ability to specify Tags for more `m.def(`
overloads.

Test Plan:
- new test

[ghstack-poisoned]

zou3519 committed:
Update on "Grandfather in c10d_functional ops to pt2_compliant"


This PR also adds the ability to specify Tags for more `m.def(`
overloads.

Test Plan:
- new test

[ghstack-poisoned]

clee2000 commented:
@pytorchbot revert -m "something in the stack broke distributed and inductor, pretty sure its this one" -c ignoredsignal 
https://hud.pytorch.org/pytorch/pytorch/commit/1d4d5e4319a5ddacdb4e0d1ac944bbb63921fdb1
https://github.com/pytorch/pytorch/actions/runs/6777904550/job/18423744447
```

2023-11-07T00:57:24.6823573Z =========================== short test summary info ============================
2023-11-07T00:57:24.6825186Z FAILED [3.3070s] distributed/test_inductor_collectives.py::TestCollectivesMultiProc::test_c10d_functional_tagged_pt2_compliant - RuntimeError: Process 0 exited with error code 10 and exception:
2023-11-07T00:57:24.6826494Z Traceback (most recent call last):
2023-11-07T00:57:24.6827558Z   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/testing/_internal/common_distributed.py", line 658, in run_test
2023-11-07T00:57:24.6828551Z     getattr(self, test_name)()
2023-11-07T00:57:24.6829571Z   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/testing/_internal/common_distributed.py", line 544, in wrapper
2023-11-07T00:57:24.6830619Z     fn()
2023-11-07T00:57:24.6831575Z   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/testing/_internal/common_utils.py", line 2558, in wrapper
2023-11-07T00:57:24.6832531Z     method(*args, **kwargs)
2023-11-07T00:57:24.6833519Z   File "/var/lib/jenkins/workspace/test/distributed/test_inductor_collectives.py", line 89, in test_c10d_functional_tagged_pt2_compliant
2023-11-07T00:57:24.6834651Z     self.assertIn(torch.Tag.pt2_compliant_tag, op.tags)
2023-11-07T00:57:24.6835462Z   File "/opt/conda/envs/py_3.10/lib/python3.10/unittest/case.py", line 1112, in assertIn
2023-11-07T00:57:24.6836266Z     self.fail(self._formatMessage(msg, standardMsg))
2023-11-07T00:57:24.6837039Z   File "/opt/conda/envs/py_3.10/lib/python3.10/unittest/case.py", line 675, in fail
2023-11-07T00:57:24.6837738Z     raise self.failureException(msg)
2023-11-07T00:57:24.6838366Z AssertionError: <Tag.pt2_compliant_tag: 8> not found in []
2023-11-07T00:57:24.6838787Z 
2023-11-07T00:57:24.6839075Z To execute this test, run the following from the base repo dir:
2023-11-07T00:57:24.6840112Z      python test/distributed/test_inductor_collectives.py -k test_c10d_functional_tagged_pt2_compliant
2023-11-07T00:57:24.6840771Z 
2023-11-07T00:57:24.6841121Z This message can be suppressed by setting PYTORCH_PRINT_REPRO_ON_FAILURE=0
2023-11-07T00:57:24.6841659Z 
2023-11-07T00:57:24.6841873Z Process 1 exited with error code 10 and exception:
2023-11-07T00:57:24.6842415Z Traceback (most recent call last):
2023-11-07T00:57:24.6843495Z   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/testing/_internal/common_distributed.py", line 658, in run_test
2023-11-07T00:57:24.6844480Z     getattr(self, test_name)()
2023-11-07T00:57:24.6845505Z   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/testing/_internal/common_distributed.py", line 544, in wrapper
2023-11-07T00:57:24.6846512Z     fn()
2023-11-07T00:57:24.6847414Z   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/testing/_internal/common_utils.py", line 2558, in wrapper
2023-11-07T00:57:24.6848415Z     method(*args, **kwargs)
2023-11-07T00:57:24.6849386Z   File "/var/lib/jenkins/workspace/test/distributed/test_inductor_collectives.py", line 89, in test_c10d_functional_tagged_pt2_compliant
2023-11-07T00:57:24.6850506Z     self.assertIn(torch.Tag.pt2_compliant_tag, op.tags)
2023-11-07T00:57:24.6851332Z   File "/opt/conda/envs/py_3.10/lib/python3.10/unittest/case.py", line 1112, in assertIn
2023-11-07T00:57:24.6852135Z     self.fail(self._formatMessage(msg, standardMsg))
2023-11-07T00:57:24.6852903Z   File "/opt/conda/envs/py_3.10/lib/python3.10/unittest/case.py", line 675, in fail
2023-11-07T00:57:24.6853630Z     raise self.failureException(msg)
2023-11-07T00:57:24.6854194Z AssertionError: <Tag.pt2_compliant_tag: 8> not found in []
2023-11-07T00:57:24.6854630Z 
2023-11-07T00:57:24.6854914Z To execute this test, run the following from the base repo dir:
2023-11-07T00:57:24.6855912Z      python test/distributed/test_inductor_collectives.py -k test_c10d_functional_tagged_pt2_compliant
```

zou3519 committed:
Update on "Grandfather in c10d_functional ops to pt2_compliant"


This PR also adds the ability to specify Tags for more `m.def(`
overloads.

Test Plan:
- new test

[ghstack-poisoned]
